runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+    /// Origin that will select winner(s), is either a given member or a council.
+    pub oracle: OracleType<MemberId>,
+
+    /// Contract type defines who can submit the work.
+    pub contract_type: AssuranceContractType<MemberId>,
+
+    /// Bounty creator member ID, should be None if created by a council.
+    pub creator_member_id: Option<MemberId>,
+
+    /// An mount of funding, possibly 0, provided by the creator which will be split among all other
+    /// contributors should the min funding bound not be reached. If reached, cherry is returned to
+    /// the creator. When council is creating bounty, this comes out of their budget, when a member
+    /// does it, it comes from an account.
+    pub cherry: Balance,
+
+    /// The minimum total quantity of funds, possibly 0, required for the bounty to become
+    /// available for people to work on.
+    pub min_amount: Balance,
+
+    /// Maximumm funding accepted, if this limit is reached, funding automatically is over.
+    pub max_amount: Balance,
+
+    /// Amount of stake required, possibly 0, to enter bounty as entrant.
+    pub entrant_stake: Balance,
+
+    /// Number of blocks from creation until funding is no longer possible. If not provided, then
+    /// funding is called perpetual, and it only ends when minimum amount is reached.
+    pub funding_period: Option<BlockNumber>,
+
+    /// Number of blocks from end of funding period until people can no longer submit
+    /// bounty submissions.
+    pub work_period: BlockNumber,
+
+    /// Number of block from end of work period until oracle can no longer decide winners.
+    pub judging_period: BlockNumber,
+
+    /// Funds provided by a bounty creator.
+    pub creator_funding: Balance,
+}
+
+/// Defines current bounty stage.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum BountyStage<BlockNumber> {
+    /// Bounty founding stage with starting block number.
+    Funding(BlockNumber),
+
+    /// A bounty was canceled.
+    Canceled,
+
+    /// A bounty was vetoed.
+    Vetoed,
+
+    /// A bounty funding was successful on the provided block.
+    /// The stage is set when the funding exceeded max funding amount.
+    MaxFundingReached(BlockNumber),
+}
+
+impl<BlockNumber: Default> Default for BountyStage<BlockNumber> {
+    fn default() -> Self {
+        BountyStage::Funding(Default::default())
+    }
+}
+
+/// Alias type for the Bounty.
+pub type Bounty<T> = BountyRecord<
+    BalanceOf<T>,
+    <T as frame_system::Trait>::BlockNumber,
+    <T as common::Trait>::MemberId,
+>;
+
+/// Crowdfunded bounty record.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct BountyRecord<Balance, BlockNumber, MemberId> {
+    /// Bounty creation parameters.
+    pub creation_params: BountyParameters<Balance, BlockNumber, MemberId>,
```

Comment:
Are you 100% sure that every single value in `BountyParameters` is indeed required to be stored?

As the bounty progresses through the different stages an increasing number of the values become redundant, but its not that big of a problem. Addressing that would require lifting out smaller and smaller subsets into the different stage variants.